### DISCUSSION TITLE
filterDeep review / refactoring

### DISCUSSION THREAD

**Discussion by YuriGor** [YuriGor/deepdash#106] [Ideas]
Same like in `filterDeep` we have `cloneDeep` option to let the user specify his own cloning method,
we need also to review all the deepdash methods where flat `_.clone` method used and also let user replace it.

**Comment by YuriGor** [comment]
Also check what's wrong with Lodash cloneDeep, as found by @GaborTorma in #56 
It keeps deep function values instead of turning them into empty objects as per docs.

**Comment by YuriGor** [comment]
Opened yet another issue about this in [Lodash](https://github.com/lodash/lodash/issues/4781)

**Comment by YuriGor** [comment]
Currently, we have these issues with filterDeep:
it pretends to return a filtered clone of a given object, but the quality of this clone is far from one provided by Lodash's clone/cloneDeep.
filterDeep assumes the data consists of objects, arrays, and primitives.
we need to carefully handle such cases as 
- class instances
- typed arrays
- maps and sets
- buffers and data views
- regexes
- functions
- read-only getters
- etc..
all this great job already is done in Lodash, so the first step will be to properly use clone/cloneDeep.
- we need let Lodash clone regexes, maps, etc
- to avoid deep cloning of data under the childrenPath
- I have a LOT to learn from Lodash clone method.
For example here is a small test how cone works:

```js
console.clear();

class Test {
  constructor(x) {
    console.log('constructor',x); // called once
    if(!x)throw 'XXX!!!'; // no error, clone doesn't call it
    this.a = "a";
    this.b = "b";
    this.c = "c";
    this.x = x;
  }
  test(){
    console.log('test');
  }
  get z(){
    return 'z';
  }
}

const t = new Test('x');
t.d='d';
delete t.b;
t.hi = ()=>{console.log('hi')};
const tt = _.cloneDeep(t);

console.log(t); // Test {a: "a", c: "c", x: "x", d: "d"}
console.log(tt); // Test {a: "a", c: "c", x: "x", d: "d"}
t.test(); // works
tt.test(); // works
tt.hi(); // works, but I bieleve it's a Lodash's bug, according to docs tt.hi should by {} after cloning
console.log(t.test===tt.test); // equal because of this method comes from prototype
console.log(t.hi===tt.hi); // equal, but I bieleve it's a Lodash's bug, according to docs tt.hi should by {} after cloning
console.log(tt.z); // getters are functions in prototype, so it works
```

**Comment by muuvmuuv** [comment]
I would also like to add a new option to exclude the children from the output when it is empty. Example: https://codepen.io/yurigor/pen/GaKvNm?editors=0010 (parent 1.2 should not have children because it is empty)

**Comment by YuriGor** [comment]
you probably posted a wrong link - your example is just template without filtering done.
But the option you are asking about is already there - read about "keepIfEmpty" ones
https://deepdash.io/#filterdeep

**Comment by muuvmuuv** [comment]
Wow... I forgot to fork it, sorry. Yes, but only if I do not use `childrenPath`: https://codepen.io/muuvmuuv/pen/VweJLPe?editors=0011

**Comment by YuriGor** [comment]
Ok got it, will think about this. 
It's not clear what to do if childrenPath is deep. not just next level property.
[like here](https://github.com/YuriGor/deepdash/blob/c7796e30b0f17ed1e32ca7e0f6d7c510414c9cfc/test/filterTree.test.js#L210)
if this option is true - then we need to remove only property by the last key of children's path, not the whole path.
